avm/src/daemon/daemon.rs

```diff
@@ -173,43 +202,86 @@ pub async fn start(
-      if vms.len() != cluster_size {
-        cluster_size = vms.len();
-        let ips = vms
-          .iter()
-          .map(|vm| vm.private_ip_addr.to_string())
-          .collect();
-        let lrh = LogRendezvousHash::new(ips);
-        leader_ip = lrh.get_leader_id().to_string();
+    let dns = DNS::new(&domain);
+    match (dns, self_ip) {
+      (Ok(dns), Ok(self_ip)) => {
+        loop {
+          sleep(period).await;
+          let mut vms_err: Option<String> = None;
+          let vms = dns.get_vms(&cluster_id).await.unwrap_or_else(|err| {
+            vms_err = Some(format!("{}", err));
+            return Vec::new();
+          });
+          if let Some(err) = vms_err {
+            error!(ErrorType::NoDnsVms, "{}", err).await;
+          };
+          // triggered the first time since cluster_size == 0
+          // and every time the cluster changes size
+          if vms.len() != cluster_size {
+            cluster_size = vms.len();
+            let ips = vms
+              .iter()
+              .map(|vm| vm.private_ip_addr.to_string())
+              .collect();
+            let lrh = LogRendezvousHash::new(ips);
+            leader_ip = lrh.get_leader_id().to_string();
+          }
+          if leader_ip == self_ip {
+            let mut factor_err: Option<String> = None;
+            let factor = post_v1_stats(&cluster_id, &deploy_token)
+              .await
+              .unwrap_or_else(|err| {
+                factor_err = Some(format!("{}", err));
+                return "1".to_string();
+              });
+            if let Some(err) = factor_err {
+              error!(ErrorType::PostStats, "{}", err).await;
+            };
+            println!(
+              "VM stats sent for cluster {} of size {}. Cluster factor: {}.",
+              cluster_id,
+              vms.len(),
+              factor
+            );
+            if factor != "1" {
+              post_v1_scale(&cluster_id, &agzb64, &deploy_token, &factor).await;
+            }
+          }
+        }
       }
-      if leader_ip == self_ip {
-        let factor = post_v1_stats(&cluster_id, &deploy_token).await;
-        println!(
-          "VM stats sent for cluster {} of size {}. Cluster factor: {}.",
-          cluster_id,
-          vms.len(),
-          factor
+      (Err(dns_err), Ok(_self_ip)) => {
+        error!(ErrorType::NoDns, "DNS error: {}", dns_err).await;
+        panic!("DNS error: {}", dns_err);
+      }
+      (Ok(_dns), Err(self_ip_err)) => {
+        error!(ErrorType::NoPrivateIp, "Private ip error: {}", self_ip_err).await;
+        panic!("Private ip error: {}", self_ip_err);
+      }
+      (Err(dns_err), Err(self_ip_err)) => {
+        error!(
+          ErrorType::NoDnsPrivateIp,
+          "DNS error: {} and Private ip error: {}", dns_err, self_ip_err
+        )
+        .await;
+        panic!(
+          "DNS error: {} and Private ip error: {}",
+          dns_err, self_ip_err
```

Comment:
I think we get rid of the `NoDnsPrivateIp` error and the complex match statement and simply do:
```
if let Err(dns_err) = dns {
  error!(ErrorType::NoDns, "DNS error: {}", dns_err).await;
        panic!("DNS error: {}", dns_err);
}
if let Err(ip_err) = _self_ip {
  error!(ErrorType:: NoPrivateIp, "Private ip error: {}", ip_err).await;
        panic!("DNS error: {}", ip_err);
}
```

Reply:
I tried something like that but was having troubles unwrapping inside the `loop`, errors like `value moved here, in previous iteration of loop`. Also tried with `unwrap_or_else` but had troubles with the returns since we need the closure to be `async` and having errors like `expected struct `DNS`, found opaque type`.

So not know if you have any other idea that I'm missing. One thing that I can do for sure is remove one of the error branches.

Reply:
Got it. That sucks but makes sense. Let's remove one branch and `NoDnsPrivateIp` then

Reply:
Wait I might thought a work around, let me check

Reply:
So this doesn't work?
```rust
if let (Ok(dns), Ok(self_ip)) = (dns, self_ip) {
  ...
} else if let Err(dns_err) = dns {
  ...
} else if let Err(ip_err) = self_ip {
  ...
}
```

Reply:
no, got 
```
error[E0382]: use of moved value
   --> src/daemon/daemon.rs:264:23
    |
213 |     let self_ip = get_private_ip().await;
    |         ------- move occurs because `self_ip` has type `std::result::Result<std::string::String, Box<dyn StdError + std::marker::Send + std::marker::Sync>>`, which does not implement the `Copy` trait
...
217 |     if let (Ok(dns), Ok(self_ip)) = (dns, self_ip) {
    |                                           ------- value moved here
...
264 |     } else if let Err(self_ip_err) = self_ip {
    |                       ^^^^^^^^^^^ value used here after move
```

Reply:
Manage to remove the `self_ip` match with an `unwrap_or_else`. Could not do the same for the `DNS` since I need to return a DNS struct and did not know how to initialize an `AsyncResolver`.
https://github.com/alantech/alan/pull/468/files?file-filters%5B%5D=.rs#diff-583f2380a4d780cb04bd93e3492a6f2828052a50e23460f3a2c8d89cf3ade79fR197-R281